src/vsr/grid_scrubber.zig

```diff
@@ -0,0 +1,481 @@
+                        "(address={} checksum={x:0>32})", .{
+                        scrubber.superblock.replica_index.?,
+                        read.read.address,
+                        read.read.checksum,
+                    });
+                }
+            }
+
+            read.ignore = read.ignore or result == .valid;
+            read.done = true;
+            scrubber.reads_busy.remove(read);
+            scrubber.reads_done.push(read);
+        }
+
+        pub fn read_reclaim(scrubber: *GridScrubber) ?struct {
+            read: *Read,
+            repair: ?BlockId,
+        } {
+            const read = scrubber.reads_done.pop() orelse return null;
+            assert(read.done);
+
+            return .{
+                .read = read,
+                .repair = if (read.ignore) null else .{
+                    .block_address = read.read.address,
+                    .block_checksum = read.read.checksum,
+                    .block_type = read.block_type.?,
+                },
+            };
+        }
+
+        fn tour_next(scrubber: *GridScrubber) ?BlockId {
+            const tour = &scrubber.tour;
+            if (tour.* == .init) {
+                tour.* = .{ .table_index = .{ .tables = ForestTableIterator{} } };
+            }
+
+            if (tour.* == .table_data) {
+                const index_block = tour.table_data.index_block orelse {
+                    // The table index is `null` if:
+                    // - It was corrupt when we just scrubbed it.
+                    // - Or `grid_scrubber_reads > 1`.
+                    // Keep trying until either we find it, or a checkpoint removes it.
+                    // (See read_next_callback() for more detail.)
+                    return .{
+                        .block_checksum = tour.table_data.index_checksum,
+                        .block_address = tour.table_data.index_address,
+                        .block_type = .index,
+                    };
+                };
+
+                const index_schema = schema.TableIndex.from(index_block);
+                const data_block_index = tour.table_data.data_block_index;
+                if (data_block_index <
+                    index_schema.data_blocks_used(scrubber.tour_index_block))
+                {
+                    tour.table_data.data_block_index += 1;
+
+                    const data_block_addresses =
+                        index_schema.data_addresses_used(scrubber.tour_index_block);
+                    const data_block_checksums =
+                        index_schema.data_checksums_used(scrubber.tour_index_block);
+                    return .{
+                        .block_checksum = data_block_checksums[data_block_index].value,
+                        .block_address = data_block_addresses[data_block_index],
+                        .block_type = .data,
+                    };
+                } else {
+                    const tables = tour.table_data.tables;
```

Comment:
assert that `data_block_index == index_schema.data_blocks_used(scrubber.tour_index_block)`